File: Python/GraphAlgorithms/GraphAlgorithms.py
from typing import List, Union

INF = float('inf')

# ===================================== EDGE CLASS =====================================
class Edge:
    vertex: int
    weight: int


    def __init__(self, v: int, w: int):
        self.vertex = v
        self.weight = w

# ===================================== GRAPH CLASS =====================================

class Graph():
    vertex_count: int           # number of vertices
    isDirected: bool            # wherther a graph is directed or not
    vertices: List[List[Edge]]  # the structure of the graph


    # --------------------------------- CLASS MANIPULATION ---------------------------------
    # constructor
    def __init__(self, vertexes: Union[int, dict[List[List[int]]], dict[List[int]]], isDirected: bool = False) -> None:
        self.isDirected = isDirected
        if type(vertexes) == int:
            self.vertex_count = vertexes
            self.vertices = [[] for i in range(vertexes)]
        else:
            self.vertex_count = len(vertexes)
            self.build(vertexes)


    # overrides string to print the class
    def __str__(self) -> str:
        s = "VERTEX -> (VERTEX|WEIGHT)\n"
        for i in range(self.vertex_count):
            s += str(i) + " -> "
            for edge in self.vertices[i]:
                s += "(" + str(edge.vertex) + "|" + str(edge.weight) + "), "
            s += "\n"
        
        return s


    # --------------------------------- ADD/REMOVE EDGE ---------------------------------
    def add_edge(self, a: int, b: int, weight: int = 1) -> None:
        self.vertices[a].append(Edge(b, weight))
        if not self.isDirected: # reverse the direction
            self.vertices[b].append(Edge(a, weight))


    def delete_edge(self, a: int, b: int) -> None:
        for edge in self.vertices[a]:
            if edge.vertex == b: 
                self.vertices[a].remove(edge)
        
        if not self.isDirected:
            for edge in self.vertices[b]:
                if edge.vertex == a: 
                    self.vertices[b].remove(edge)


    # --------------------------------- BUILD GRAPH ---------------------------------
    def build(self, graph: Union[dict[List[List[int]]], dict[List[int]]]) -> None:
        self.vertices = [[] for _ in range(self.vertex_count)]
        for vertex in graph:
            for edge in graph[vertex]:
                try:
                    self.add_edge(vertex, edge[0], edge[1]) # weighted edge
                except:
                    self.add_edge(vertex, edge) # unweighted edge   


    # --------------------------------- FIND PATH IN TREE ---------------------------------
    def Path(self, start: int, finish: int, tree: List[int]) -> List[int]:
        path: List[int] = []

        if tree[finish] != -1:  # start != finish
            while finish != start:
                path.append(finish)
                finish = tree[finish]
            path.append(start)
            path.reverse()

        return path


    # --------------------------------- BREADTH FIRST SEARCH ---------------------------------
    def BFS_tree(self, start: int) -> List[int]:
        visited: List[bool] = [False for _ in range(self.vertex_count)]
        tree: List[int] = [-1 for _ in range(self.vertex_count)]
        queue: List[int] = []
        v: int # current vertex

        visited[start] = True
        queue.append(start) # put the start in the queue

        while queue:
            v = queue.pop(0)

            for edge in self.vertices[v]:
                if not visited[edge.vertex]:
                    visited[edge.vertex] = True
                    tree[edge.vertex] = v
                    queue.append(edge.vertex)

        return tree
    

    def BFS(self, start: int, finish: int) -> List[int]:
        tree: List[int] = self.BFS_tree(start) # get the tree from BFS algorithm
        return self.Path(start, finish, tree)


    # --------------------------------- DEPTH FIRST SEARCH ---------------------------------
    def DFS_tree(self, start: int) -> List[int]:
        visited: List[bool] = [False for _ in range(self.vertex_count)]
        tree: List[int] = [-1 for _ in range(self.vertex_count)]
        stack: List[int] = []
        v: int # current vertex

        stack.append(start) # put the start in the stack

        while stack:
            v = stack.pop() # take the top item on stack

            if not visited[v]:
                visited[v] = True

                for edge in self.vertices[v]:
                    stack.append(edge.vertex)
                    if not visited[edge.vertex]: tree[edge.vertex] = v
            
        return tree


    def DFS(self, start: int, finish: int) -> List[int]:
        tree: List[int] = self.DFS_tree(start) # get the tree from DFS algorithm
        return self.Path(start, finish, tree)


    # --------------------------------- DIJKSTRA SEARCH ---------------------------------
    def Dijkstra_tree(self, start: int) -> List[int]:
        distance: List[int] = [INF for _ in range(self.vertex_count)] # distance from start node
        tree: List[int] = [-1 for _ in range(self.vertex_count)]
        queue: List[int] = []
        v: int # current vertex

        distance[start] = 0
        queue.append(start)

        while queue:
            v = queue.pop(0)

            for edge in self.vertices[v]:
                if distance[v] + edge.weight < distance[edge.vertex]:
                    distance[edge.vertex] = distance[v] + edge.weight
                    tree[edge.vertex] = v
                    queue.append(edge.vertex)
        
        return tree


    def Dijkstra(self, start: int, finish: int) -> List[int]:
        tree: List[int] = self.Dijkstra_tree(start) # get the tree from Dijkstra algorithm
        return self.Path(start, finish, tree)
    
    # --------------------------------- A* SEARCH ---------------------------------
    def AStar_tree(self, start: int) -> List[int]:
        # Nah bro i give up. (╯°□°）╯︵ ┻━┻
        # I've been staring at the screen for 2 hours watching an indian guy explain heuristic values. ༼ つ ◕_◕ ༽つ
        pass

File: C++/GraphAlgorithms/GraphAlgorithms.cpp
#include <vector>
#include <unordered_set>
#include <stack>
#include <queue>
#include <algorithm>
#include "GraphAlgorithms.hpp"


// ---------------------------- CLASS MANIPULATION ----------------------------
// create a graph by the number of vertices
Graph::Graph(int vertexes, bool directed) {
    vertexCount = vertexes;
    isDirected = directed;
    graph = std::vector<std::unordered_set<Edge, EdgeHash>>(vertexCount, std::unordered_set<Edge, EdgeHash>());
}

// create an unweighted graph from HashMap
Graph::Graph(std::vector<std::vector<int>> graph, bool directed){
    vertexCount = graph.size();
    isDirected = directed;
    this->graph = std::vector<std::unordered_set<Edge, EdgeHash>>(vertexCount, std::unordered_set<Edge, EdgeHash>());
    
    for(int i = 0; i < graph.size(); i++){
        for(int edge : graph[i]){
            if(edge+1 > vertexCount) throw std::invalid_argument("Vertex " + std::to_string(edge) + " doesn't exist");
            addEdge(i, edge);
            if(!directed) addEdge(edge, i);
        }
    }
}

// create a weighted graph from HashMap
Graph::Graph(std::vector<std::vector<std::vector<int>>> graph, bool directed){
    vertexCount = graph.size();
    isDirected = directed;
    this->graph = std::vector<std::unordered_set<Edge, EdgeHash>>(vertexCount, std::unordered_set<Edge, EdgeHash>());
    
    for(int i = 0; i < graph.size(); i++){
        for(std::vector<int> edge : graph[i]){
            if(edge[0]+1 > vertexCount) throw std::invalid_argument("Vertex " + std::to_string(edge[0]) + " doesn't exist");
            addEdge(i, edge[0], edge[1]);
            if(!directed) addEdge(edge[0], i, edge[1]);
        }
    }
}


// ---------------------------- ADD/REMOVE EDGE ----------------------------
void Graph::addEdge(int a, int b, int weigth) {
    graph[a].insert(Edge(b, weigth));
    if(!isDirected) graph[b].insert(Edge(a, weigth));
}

void Graph::removeEdge(int a, int b) {
    graph[a].erase(Edge(b, 1));
    if(!isDirected) graph[b].erase(Edge(a, 1));
}

// ---------------------------- FIND PATH IN TREE ----------------------------
std::vector<int> Graph::path(int start, int finish, std::vector<int> tree){
    std::vector<int> path;

    // there is no path
    if(tree[finish] == -1) return path;

    while(start != finish) {
        path.push_back(finish);
        finish = tree[finish];
    }

    path.push_back(start);
    std::reverse(path.begin(), path.end());
    return path;
}

// ---------------------------- BREADTH FIRST SEARCH ----------------------------
std::vector<int> Graph::BFSTree(int start) {
    std::vector<bool> visited(graph.size(), false);
    std::vector<int> tree(graph.size(), -1);
    std::queue<int> q;

    visited[start] = true;
    q.push(start);

    while(q.size() > 0){
        int v = q.front();
        q.pop();

        for(Edge edge : graph[v]){
            if(!visited[edge]){
                visited[edge] = true;
                tree[edge] = v;
                q.push(edge);
            }
        }
    }

    return tree;
}

std::vector<int> Graph::BFS(int start, int finish) {
    std::vector<int> tree = BFSTree(start);
    return path(start, finish, tree);
}

// ---------------------------- DEPTH FIRST SEARCH ----------------------------
std::vector<int> Graph::DFSTree(int start) {
    std::vector<bool> visited(graph.size(), false);
    std::vector<int> tree(graph.size(), -1);
    std::stack<int> s;

    visited[start] = true;
    s.push(start);

    while(s.size() > 0){
        int v = s.top();
        s.pop();

        for(Edge edge : graph[v]){
            if(!visited[edge]){
                visited[edge] = true;
                tree[edge] = v;
                s.push(edge);
            }
        }
    }

    return tree;
}

std::vector<int> Graph::DFS(int start, int finish) {
    std::vector<int> tree = DFSTree(start);
    return path(start, finish, tree);
}

// ---------------------------- DIJKSTRA'S ALGORITHM ----------------------------
std::vector<int> Graph::DijkstraTree(int start){
    std::vector<int> distance(graph.size(), INT_MAX);
    std::vector<int> tree(graph.size(), -1);
    std::priority_queue<int> pq;

    distance[start] = 0;
    pq.push(start);

    while(pq.size() > 0){
        int v = pq.top();
        pq.pop();

        for(Edge edge : graph[v]){
            if(distance[v] + edge.weigth < distance[edge]){
                distance[edge] = distance[v] + edge.weigth;
                tree[edge] = v;
                pq.push(edge);
            }
        }
    }

    return tree;
}

std::vector<int> Graph::Dijkstra(int start, int finish){
    std::vector<int> tree = DijkstraTree(start);
    return path(start, finish, tree);
}
